File: python/300-399/377_combination_sum_4.py
import sys
from typing import List


class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0 for i in range(target + 1)]
        dp[0] = 1

        for i in range(1, target + 1):
            dp[i] = sum(dp[i - num] for num in nums if i - num >= 0)

        return dp[-1]


solution = Solution()
print(solution.combinationSum4(nums = [1,2,3], target = 4))

File: c++/300-399/377_combination_sum4.cpp
#include <iostream>
#include <vector>
#include <climits>

using std::vector;

class Solution {
public:
    int combinationSum4(const vector<int>& nums, int target) {
        vector dp (target+1, 0);
        dp[0] = 1;

        for (int i = 0; i <= target; i++) {
            for (auto& n: nums) {
                if (i >= n) {
                    dp[i] += dp[i-n];
                }
            }
        }

        return dp[target];
    }
};

int main() {
    auto s = Solution();
    std::cout << s.combinationSum4({9}, 3) << std::endl;
}
